clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1799,4 +1801,42 @@ void handleFailure(Throwable throwable) {
 
         return new DescribeReplicaLogDirResult(new HashMap<TopicPartitionReplica, KafkaFuture<ReplicaLogDirInfo>>(futures));
     }
+
+    public CreatePartitionsResult createPartitions(Map<String, NewPartition> newPartitions, final CreatePartitionsOptions options) {
+        final Map<String, KafkaFutureImpl<Void>> futures = new HashMap<>(newPartitions.size());
+        for (String topic: newPartitions.keySet()) {
+            futures.put(topic, new KafkaFutureImpl<Void>());
+        }
+        final Map<String, NewPartition> requestMap = new HashMap<>(newPartitions);
+
+        final long now = time.milliseconds();
+        runnable.call(new Call("createPartitions", calcDeadlineMs(now, options.timeoutMs()),
+                new ControllerNodeProvider()) {
+
+            @Override
+            public AbstractRequest.Builder createRequest(int timeoutMs) {
+                return new CreatePartitionsRequest.Builder(requestMap, timeoutMs, options.shouldValidateOnly());
+            }
+
+            @Override
+            public void handleResponse(AbstractResponse abstractResponse) {
+                CreatePartitionsResponse response = (CreatePartitionsResponse) abstractResponse;
+                for (Map.Entry<String, ApiError> result : response.errors().entrySet()) {
+                    KafkaFutureImpl<Void> future = futures.get(result.getKey());
+                    if (result.getValue().isSuccess()) {
+                        future.complete(null);
+                    } else {
+                        future.completeExceptionally(result.getValue().exception());
+                    }
+                }
+            }
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                completeAllExceptionally(futures.values(), throwable);
+            }
+        }, now);
+        return new CreatePartitionsResult(new HashMap<String, KafkaFuture<Void>>(futures));
```

Comment:
Another HashMap is created.
Is it possible not to ?

Reply:
@tedyu without the copy `futures` would be simultaneously accessible to both the `Call` and the `CreatePartitionsResult`. I will admit I didn't think very hard about it because the same pattern is used by all the other methods in `KafkaAdminClient`. 

As it happens both `Call` and `CreatePartitionsResult` only call `get()` and `values()` on the `HashMap`, so if you're willing it make assumptions about the internals of `HashMap`, you could make the argument that it would be safe for them to share the instance. But making assumptions and correct multi-threaded code have plenty of ugly history, so I think making a copy is warranted on the grounds of defensive programming.

Reply:
When we get here, futures Map is already filled.
Calling get() and values() does not involve the internals of {{HashMap}} when we consider thread-safety.
What do you think ?

Reply:
See KAFKA-5911

Reply:
I agree that it's safe to avoid the copy here. By making the copy, we are iterating from a different thread anyway, so it doesn't add any safety.